core/domain/topic_jobs_one_off.py

```diff
@@ -92,3 +94,68 @@ def reduce(key, values):
                 sum(ast.literal_eval(v) for v in values))])
         else:
             yield (key, values)
+
+
+class RemoveDeletedUncategorizedSkillsOneOffJob(
+        jobs.BaseMapReduceOneOffJobManager):
+    """One-off job to remove deleted uncategorized skills linked to a topic."""
+
+    _DELETED_KEY = 'topic_deleted'
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [topic_models.TopicModel]
+
+    @staticmethod
+    def map(item):
+        if item.deleted:
+            yield (RemoveDeletedUncategorizedSkillsOneOffJob._DELETED_KEY, 1)
+            return
+
+        # Note: the read will bring the topic up to the newest version.
+        topic = topic_fetchers.get_topic_by_id(item.id)
+        skill_ids_to_be_removed_from_subtopic = []
+        all_skill_ids_to_be_removed = []
+        commit_cmds = []
+        for subtopic in topic.get_all_subtopics():
+            local_skill_models = skill_models.SkillModel.get_multi(
+                subtopic['skill_ids'])
+            for skill_id, skill_model in python_utils.ZIP(
+                    subtopic['skill_ids'], local_skill_models):
+                if skill_model is None:
+                    commit_cmds.append(topic_domain.TopicChange({
+                        'cmd': topic_domain.CMD_REMOVE_SKILL_ID_FROM_SUBTOPIC,
+                        'skill_id': skill_id,
+                        'subtopic_id': subtopic['id']
+                    }))
+                    skill_ids_to_be_removed_from_subtopic.append(skill_id)
+
+        local_skill_models = skill_models.SkillModel.get_multi(
+            topic.get_all_skill_ids())
+        list_of_all_skill_ids = (
+            topic.uncategorized_skill_ids +
+            skill_ids_to_be_removed_from_subtopic)
```

Comment:
Why are you processing skill_ids_to_be_removed_from_subtopic twice? (it seems like you're making multiple commit_cmds for them -- this one and the one above?)

Reply:
Let me try to explain what I'm doing here:
I'm unassigning skills assigned to subtopics using the `CMD_REMOVE_SKILL_ID_FROM_SUBTOPIC` commit cmd. This adds the skill ids back to uncategorized skill ids in the topic model. Then I remove uncategorized skill ids from the topic model using `CMD_REMOVE_UNCATEGORIZED_SKILL_ID` (this will include ones that were unassigned from subtopics). This seemed like the right way to cleanly get rid of the skills ids associated with topics and subtopics.

Reply:
The change list that will be executed will be in this order:
1. Remove skill ids from subtopic for skills that are deleted and make them uncategorized skills in the topic.
2. Remove uncategorized skills (for skills which are deleted) from the topic. These uncategorized skills will include the skills removed from subtopic in (1).